posts/2022-07-12-components-with-dynamic-attributes.md

```diff
@@ -0,0 +1,302 @@
+
+Our weekly planning task board should look something similar to this when we open our browser:
+
+
+![board plan output](/images/board-plan-output.png)
+
+
+## 2.Problem: Reusing components with different attributes
+
+  1. We want the cards in each column to have a different color from one another . For example, cards in the "work" column should be blue, cards in the "house" column should be green, and cards in the "school" column should be yellow.
+
+  2. We should be able to use the same card/1 function component we defined earlier, while still ensuring that cards can be a different color in the different columns.
+
+
+We could solve the first problem like this:
+
+```elixir
+    <div class="row">
+      <div class="column">
+        <h3>Work</h3>
+        <%= for card <- @work_cards do %> 
+         <div>
+          <div class="column bg-blue">
+            <%= card.task %>
+          </div>
+         </div>
+        <% end %>
+       </div>
+       <div class="column">
+         <h3>House</h3>
+         <%= for card <- @house_cards do %> 
+           <div>
+            <div class="column bg-green">
+             <%= card.task %>
+            </div>
+          </div>
+         <% end %>
+       </div>
+       <div class="column">
+         <h3>School</h3>
+         <%= for card <- @school_cards do %> 
+          <div>
+           <div class="column bg-yellow">
+            <%= card.task %>
+           </div>
+         </div>
+         <% end %>
+        </div>
+      </div>
+
+```
+
+We have solved our first problem , its working but this code has some shortcomings. Personally, its redundant to write the same markup over and over again and this is the reason why we opted to use function components in the first place. However, our original implementation doesn't allow us to control the color of the cards in each column.
+
+
+How can we exactly solve this problem?
+
+
+## 3. Solution: Dynamic Component Attributes with `assigns_to_attributes/2`
+
+Luckily, Phoenix LiveView v0.16.0 introduced the [assigns_to_attribute/2](https://hexdocs.pm/phoenix_live_view/0.16.0/Phoenix.LiveView.Helpers.html#assigns_to_attributes/2) function. 
+
+_This function is useful for transforming caller assigns into dynamic attributes while stripping reserved keys from the result._
+
+`assigns_to_attribute/2` takes in assigns as the first argument and a list of assign’s keys that are to be excluded as the optional second argument.Then it returns a filtered list of keywords for use as HTML attributes.
+
+Now that we are assured we can transform the assigns passed to `card/1` into attributes, let’s go ahead and add an assigns of `class` to our call to `card/1`. We'll give that assigns a value of `"bg-*"`. “bg-*” represents background color with a CSS added property.
+
+
+```elixir
+<h3>Work</h3>
+<%= for card <- @work_cards do %> 
+   <.card card={card} class={"bg-blue"}/>
+ <% end %>
+.
```

Comment:
Remove the lines that just contain `.`. If you need an ellipsis, use an HTML comment like this:

```HTML
<!--- ... -->
```